# Спектральное разложение

Спектральное разложение матрицы — это представление **квадратной** матрицы ${\displaystyle A}$ в виде произведения трёх матриц, ${\displaystyle A=V\Lambda V^{-1}}$, где 

- ${\displaystyle V}$ — матрица, столбцы которой являются ортонормированными собственными векторами матрицы ${\displaystyle A}$, 
- ${\displaystyle \Lambda }$  — диагональная матрица с соответствующими собственными значениями на главной диагонали, 
- ${\displaystyle V^{-1}}$ — матрица, обратная матрице ${\displaystyle V}$.

Обозначение из лекций: $X = S^T \cdot  D \cdot S$

In [1]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

A = np.array([[1.0, 2.0, -1.0],
              [2.0, 2.0, 5.0],
              [-1.0, 5.0, -3.0]])
A

array([[ 1.,  2., -1.],
       [ 2.,  2.,  5.],
       [-1.,  5., -3.]])

In [2]:
v, w = np.linalg.eigh(A)

In [3]:
v

array([-6.58,  1.15,  5.43])

In [4]:
w

array([[-0.25,  0.93, -0.27],
       [ 0.53, -0.11, -0.84],
       [-0.81, -0.35, -0.47]])

In [5]:
D = np.diag(v)

In [6]:
D

array([[-6.58,  0.  ,  0.  ],
       [ 0.  ,  1.15,  0.  ],
       [ 0.  ,  0.  ,  5.43]])

In [7]:
S = w.T

In [8]:
S

array([[-0.25,  0.53, -0.81],
       [ 0.93, -0.11, -0.35],
       [-0.27, -0.84, -0.47]])

In [9]:
S.T

array([[-0.25,  0.93, -0.27],
       [ 0.53, -0.11, -0.84],
       [-0.81, -0.35, -0.47]])

In [10]:
np.dot(S.T, S)

array([[ 1.,  0.,  0.],
       [ 0.,  1.,  0.],
       [ 0.,  0.,  1.]])

In [11]:
np.dot(np.dot(S.T, D), S)

array([[ 1.,  2., -1.],
       [ 2.,  2.,  5.],
       [-1.,  5., -3.]])

## Квадратичные формы
$$P(x, y, z) = x^2 + 2 y^2 - 3 z^2 + 4 xy + 10 yz - 2 xz$$

$$
P(x, y, z) =
\left(
\begin{array}{ccc}
x & y & z
\end{array}
\right)
\left(
\begin{array}{ccc}
1 & 2 & -1 \\
2 & 2 & 5 \\
-1 & 5 & -3 \end{array}
\right)
\left(
\begin{array}{c}
x \\
y \\
z \end{array}
\right)
$$

In [12]:
A = np.array([[1.0, 2.0, -1.0],
              [2.0, 2.0, 5.0],
              [-1.0, 5.0, -3.0]])
v, W = np.linalg.eigh(A)
D = np.diag(v)
S = W.T

print ('D =\n', D)
print ('S=\n', S)

D =
 [[-6.58  0.    0.  ]
 [ 0.    1.15  0.  ]
 [ 0.    0.    5.43]]
S=
 [[-0.25  0.53 -0.81]
 [ 0.93 -0.11 -0.35]
 [-0.27 -0.84 -0.47]]


Таким образом,
$$P(x, y, z) = -6.58 \cdot \tilde{x}^2 + 1.15 \cdot \tilde{y}^2 + 5.43 \cdot \tilde{z}^2,$$
где
$$
\left(
\begin{array}{c}
\tilde{x} \\
\tilde{y} \\
\tilde{z} \end{array}
\right) =
S
\left(
\begin{array}{c}
x \\
y \\
z \end{array}
\right),
$$ т.е.
$$
\tilde{x} = -0.25 x + 0.53 y - 0.81 z,
$$
$$
\tilde{y} = 0.93 x - 0.11 y - 0.35 z,
$$
$$
\tilde{z} = -0.27 x - 0.84 y - 0.47 z.
$$

# Сингулярное разложение

Сингулярным разложением матрицы ${\displaystyle M}$ порядка $m\times n$ является разложение вида $M=U\Sigma V^{*}$, где 

- $\Sigma$  — матрица размера $m\times n$ с неотрицательными элементами, у которой элементы, лежащие на главной диагонали — это сингулярные числа (а все элементы, не лежащие на главной диагонали, являются нулевыми), 
- а матрицы $U$ (порядка $m$) и $V$ (порядка $n$) — это две унитарные матрицы, состоящие из левых и правых сингулярных векторов соответственно (а $V^*$ — это сопряжённо-транспонированная матрица к $V$).


Обозначение из лекций: $X = U \cdot  D \cdot V$

In [13]:
X = np.array([[1, 2, 0],
              [0, 0, 5],
              [3, -4, 2],
              [1, 6, 5],
              [0, 1, 0]])

X

array([[ 1,  2,  0],
       [ 0,  0,  5],
       [ 3, -4,  2],
       [ 1,  6,  5],
       [ 0,  1,  0]])

In [14]:
U, s, W = np.linalg.svd(X)

In [15]:
U

array([[ 0.17,  0.16, -0.53, -0.8 , -0.16],
       [ 0.39, -0.53,  0.61, -0.43,  0.03],
       [-0.14, -0.82, -0.52,  0.14,  0.07],
       [ 0.89,  0.06, -0.25,  0.38, -0.06],
       [ 0.08,  0.11, -0.08, -0.11,  0.98]])

In [16]:
W

array([[ 0.07,  0.72,  0.69],
       [-0.37,  0.67, -0.65],
       [-0.93, -0.21,  0.31]])

In [17]:
V = W.T
V

array([[ 0.07, -0.37, -0.93],
       [ 0.72,  0.67, -0.21],
       [ 0.69, -0.65,  0.31]])

In [18]:
s

array([ 8.82,  6.14,  2.53])

In [19]:
D = np.zeros_like(X, dtype=float)
D

array([[ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.],
       [ 0.,  0.,  0.]])

In [20]:
D[np.diag_indices(min(X.shape))] = s
D

array([[ 8.82,  0.  ,  0.  ],
       [ 0.  ,  6.14,  0.  ],
       [ 0.  ,  0.  ,  2.53],
       [ 0.  ,  0.  ,  0.  ],
       [ 0.  ,  0.  ,  0.  ]])

In [21]:
np.dot(U.T, U)

array([[ 1.,  0., -0., -0., -0.],
       [ 0.,  1.,  0., -0.,  0.],
       [-0.,  0.,  1., -0., -0.],
       [-0., -0., -0.,  1., -0.],
       [-0.,  0., -0., -0.,  1.]])

In [22]:
np.dot(V.T, V)

array([[ 1.,  0.,  0.],
       [ 0.,  1., -0.],
       [ 0., -0.,  1.]])

In [23]:
np.dot(np.dot(U, D), V.T)

array([[ 1.,  2.,  0.],
       [ 0., -0.,  5.],
       [ 3., -4.,  2.],
       [ 1.,  6.,  5.],
       [-0.,  1.,  0.]])

# Связь между спектральным и сингулярным разложениями

In [24]:
A = np.array([[1.0, 2.0, -1.0],
              [2.0, 2.0, 5.0],
              [-1.0, 5.0, -3.0]])
A

array([[ 1.,  2., -1.],
       [ 2.,  2.,  5.],
       [-1.,  5., -3.]])

Спектральное разложение:

In [26]:
v, w = np.linalg.eigh(A)
D_sp = np.diag(v)
S_sp = w.T

print(S_sp.T, D_sp, S_sp, sep='\n\n')

[[-0.25  0.93 -0.27]
 [ 0.53 -0.11 -0.84]
 [-0.81 -0.35 -0.47]]

[[-6.58  0.    0.  ]
 [ 0.    1.15  0.  ]
 [ 0.    0.    5.43]]

[[-0.25  0.53 -0.81]
 [ 0.93 -0.11 -0.35]
 [-0.27 -0.84 -0.47]]


In [27]:
np.dot(np.dot(S_sp.T, D_sp), S_sp)

array([[ 1.,  2., -1.],
       [ 2.,  2.,  5.],
       [-1.,  5., -3.]])

Сингулярное разложение:

In [28]:
U_sin, s, W = np.linalg.svd(A)
D_sin = np.zeros_like(A, dtype=float)
D_sin[np.diag_indices(min(A.shape))] = s

V_sin = W.T

print(U_sin, D_sin, V_sin.T, sep='\n\n')

[[ 0.25 -0.27  0.93]
 [-0.53 -0.84 -0.11]
 [ 0.81 -0.47 -0.35]]

[[ 6.58  0.    0.  ]
 [ 0.    5.43  0.  ]
 [ 0.    0.    1.15]]

[[-0.25  0.53 -0.81]
 [-0.27 -0.84 -0.47]
 [ 0.93 -0.11 -0.35]]


In [29]:
np.dot(np.dot(U_sin, D_sin), V_sin.T)

array([[ 1.,  2., -1.],
       [ 2.,  2.,  5.],
       [-1.,  5., -3.]])